### PR TITLE
[flutter_tools] hot reload/restart update for asset manager change

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Do not upload all assets on initial devFS sync.  This should increase the reliability of the initial connection, even in the face of flaky devfs behavior, in addition to a moderate perf improvement.

Updates fast-start to build assets as part of the initial bundle

Requires https://github.com/flutter/engine/pull/21436
Requires https://github.com/flutter/engine/pull/21586
Requires https://github.com/flutter/engine/pull/21611

### CODE REVIEW

File: packages/flutter_tools/lib/src/devfs.dart [hunk 1]
```diff
@@ -490,13 +490,12 @@ class DevFS {
...
-        final Uri deviceUri = _fileSystem.path.toUri(_fileSystem.path.join(assetDirectory, archivePath));
-        if (deviceUri.path.startsWith(assetBuildDirPrefix)) {
-          archivePath = deviceUri.path.substring(assetBuildDirPrefix.length);
-        }
-        // Only update assets if they have been modified, or if this is the
-        // first upload of the asset bundle.
-        if (content.isModified || (bundleFirstUpload && archivePath != null)) {
+        // Only update assets if they have been modified.
+        if (content.isModified) {
+          final Uri deviceUri = _fileSystem.path.toUri(_fileSystem.path.join(assetDirectory, archivePath));
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/devfs.dart:495]: bonus perf improvement by moving this under the check

File: packages/flutter_tools/lib/src/devfs.dart [hunk 2]
```diff
@@ -472,7 +472,6 @@ class DevFS {
     String dillOutputPath,
     bool fullRestart = false,
     String projectRootPath,
-    bool skipAssets = false,
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/devfs.dart:475]: Not used

File: packages/flutter_tools/lib/src/devfs.dart [hunk 3]
```diff
@@ -484,19 +483,17 @@ class DevFS {
...
       final String assetDirectory = getAssetBuildDirectory();
       bundle.entries.forEach((String archivePath, DevFSContent content) {
-        final Uri deviceUri = _fileSystem.path.toUri(_fileSystem.path.join(assetDirectory, archivePath));
-        if (deviceUri.path.startsWith(assetBuildDirPrefix)) {
-          archivePath = deviceUri.path.substring(assetBuildDirPrefix.length);
-        }
-        // Only update assets if they have been modified, or if this is the
-        // first upload of the asset bundle.
-        if (content.isModified || (bundleFirstUpload && archivePath != null)) {
+        if (content.isModified && !bundleFirstUpload) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/devfs.dart:492]: nit: flip the sense and unindent
```dart
if (!content.isModified || bundleFirstUpload) {
  return;
}
...
```

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/devfs.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/devfs.dart]: Fixed

**Comment by dnfield** [comment]
LGTM as well.

**Comment by jonahwilliams** [comment]
Google testing passed

**Pull request merged**